p4-16/psa/psa.p4

```diff
@@ -373,6 +377,70 @@ extern ValueSet<D> {
 }
 // END:ValueSet_extern
 
+// BEGIN:Timestamp_extern
+extern Timestamp {
+    // No constructor.
+
+    // The nanoseconds, microseconds, and milliseconds values
+    // converted to or from via these methods are the same as those
+    // defined as the Java Time-Scale,
+    // documented here:
+    // https://docs.oracle.com/javase/8/docs/api/java/time/Instant.html
+
+    // As a rough description that is intended only for getting a
+    // quick idea, they are intended to be the number of those time
+    // units elapsed since midnight on January 1, 1970.  The intent is
+    // that, _within the limits of whatever time synchronization
+    // mechanisms are in use on the device_, these values might be
+    // useful in correlating event times across multiple devices in a
+    // network, e.g. in an Inband Network Telemetry data collector.
+
+    // The "_to_timestamp" methods of course allow multiple input
+    // times to return the same timestamp_t value, in the expected
+    // case that a timestampt_t has few enough bits that it wraps
+    // within the full range of the input values.
+
+    // For the same reason, the "timestamp_to_" methods can return
+    // different values depending upon the time at which they are
+    // called.  They will return the correct external time only when
+    // given as input a timestamp value t that meets the following
+    // conditions:
+
+    // If you get a timestamp value and "very soon afterwards" call
+    // one of the timestamp_to_* methods, it will return the correct
+    // value.
+
+    // If you get a timetamp value and store it for up to half of a
+    // timestamp wrap time, then call one of the timestamp_to_*
+    // methods, it will return the correct value.
+
+    // The timestamp_to_* methods should not be relied upon to return
+    // the correct values for future timestamps, nor for timestamps
+    // more than half of a wrap time old.
+
+    void timestamp_to_nanoseconds(in timestamp_t t,
+                                  out bit<64> nanoseconds);
+    void timestamp_to_microseconds(in timestamp_t t,
+                                   out bit<64> microseconds);
+    void timestamp_to_milliseconds(in timestamp_t t,
+                                   out bit<64> milliseconds);
+    void nanoseconds_to_timestamp(in bit<64> nanoseconds,
+                                  out timestampt_t t);
+    void microseconds_to_timestamp(in bit<64> microseconds,
+                                   out timestampt_t t);
+    void milliseconds_to_timestamp(in bit<64> milliseconds,
+                                   out timestampt_t t);
```

Comment:
Are these APIs really needed in the data plane? If you had just the timestamps you can get the relative delta between events, is it important in the data plane to relate them back to [m,u,n]sec?